core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -600,6 +600,15 @@ class UserService(
       case ProjectRoles.User  => SamBillingProjectPolicyNames.workspaceCreator
     }
 
+  def verifyBillingProjectAccess(projectUuid: UUID, action: SamResourceAction): Future[Boolean] =
+    for {
+      billingProject <- billingRepository.getBillingProjectById(projectUuid)
+      userHasAction <- billingProject match {
+        case Some(bp) => samDAO.userHasAction(SamResourceTypeNames.billingProject, bp.projectName.value, action, ctx)
+        case None     => Future.successful(false)
+      }
+    } yield userHasAction
+
```

Comment:
Optional: if the billing project exists, but user does not have the requested action … you could make a second call to Sam to see if the user has "read" on that BP. This would allow you to return 403 when the user has read and 404 when the user does not. But this might be overkill for this niche API; it would only be worth it if TDR wanted to distinguish between those use cases.

Reply:
Yeah, I considered that but ended up going with this approach to more closely mimic the behavior of the [Sam endpoint](https://sam.dsde-dev.broadinstitute.org/#/Resources/resourcePermissionV2) which returns false whether the resource exists or not. Thinking about it some more though, it might be good to distinguish between a 403 and a 404 here, the Sam API always returns 200, so this has already diverged from the Sam behavior

Reply:
The `registerGoogleProject` API also distinguishes between 403 and 404, and could maybe be modified to use this method